### ISSUE TITLE
vine: transfer files to free resources

### ISSUE THREAD

**Issue by btovar** [cooperative-computing-lab/cctools#3538]
Related to #3531, a question that came up yesterday at pyhep was whether taskvine could consolidate files into a fewer workers when there are not many tasks left to be run. That is, if the only service the worker is currently providing is holding a file, maybe it can transfer a file to some other worker and idle out.

I think this could be related to the idea that @colinthomas-z80 proposed, where workers ask permission to the manager to idle out. The worker could tell the manager that it wants to idle out, and the manager can release the worker after the files have been transferred.

**Comment by btovar** [comment]
By itself the manager does not seem to have enough information. It needs to get a hint from the application (e.g., the application tells the manager the expected number of tasks yet to be submitted), or from the factory (e.g., is the factory draining?), is the application in ramp-down, etc.  (Related to this, I think we should add an api call that let's the manager know the stages "ramp-up", "steady-state", "ramp-down", as it should be doing things differently in each stage: ramp-up: send as many tasks as possible to as many workers as possible;  steady-state: give preference to retrievals; ramp-down: defragment, use whole workers to avoid resource exhaustion.)

In some sense, I don't think is up to the worker to decide to idle out if it still has manger assets. It is a resource management problem for the manager.  When appropriate from the hints above, the manager may "defragment" its distributed cached, and in doing so some workers will be emptied and then may decide to idle out.

Note that this is somewhat different from workers having a disconnection deadline, as idling-out is a choice.  Further, we cannot treat disconnection deadlines and timeouts the same, as timeouts tend to be more shorter. If we treat them the same, then there may be a cascading effect where all the workers ask to be released because the only activity in the cluster is from the manager moving files around to allow previous workers to idle-out.

**Comment by btovar** [comment]
For WQ, I don't think it is quite as easy. For clusters where the batch queue is short, then yes, idling-out right away makes sense. But if the batch queue is long, for some applications it is more efficient to keep the worker idle so it can execute tasks that have not been submitted.

**Comment by colinthomas-z80** [comment]
Our conclusions today were largely the same. The manager should decide whether the worker can exit. We found that the set of decisions a worker can currently make will either transfer an unwanted large amount of data, or prevent ramp-down and use an unnecessary amount of cores. 

The discussion brought up another point about workflow declaration, which is that all of the tasks should be declared in the beginning so the manager has the full dag in view. We discussed a potential API call to signal we are done submitting tasks.

The manager already has a reference count of replicated files, and we could preemptively let go of workers that do not have unique or important data if we can determine they won't be needed for queued tasks. 

Consolidating data and letting go of further workers sounds like a good idea, but where do we choose to send the data? Sending it back to the manager could undesirably block the manager and potentially blow up the disk if we have cascading timeout workers. If we send it to another worker we should reset its timeout so that it does not in turn request to leave. 

I think this correlates to a recent desire of mine for preemptive data staging before a task is ready to run. In both of our situations we have tasks with numerous data inputs waiting until all of them are ready before transferring anything. If future dependencies can be consolidated we can let go of idle cores and help the next tasks get started at the same time. This might be critical in matching the performance of a statically compiled dag (pegasus).

**Comment by btovar** [comment]
@colinthomas-z80 Agreed, very interesting. I prefer having the  release of workers as manager decision, rather than the workers asking to be released because of their timeouts. I don't think that the timeout value should have any influence in this decision. That is, the resource utilization efficiency should be the same if I set the worker timeout to five minutes or one hour. Thus, I don't think workers should ask for permission to be released. Workers can tell the manager when they are going away, however, like if submitted via slurm with a hard deadline, and the manager can plan to move those files away.

One of the strengths of taskvine is to have dynamic workflows, so whatever solution should be available even if not all tasks are declared up-front. So, either the manager is told that all tasks have been submitted, or it is told which stage the workflow is in (as mentioned in the comment above). Another complication is that it is not enough to know that we are in the final stages of a workflow. Logically that might be the case, but if it coincides with workers being evicted and files are lost, then we need all the workers that we can get.

Where to consolidate files is an interesting policy decision. Maybe we prefer workers that just connected as they are less likely to be evicted, or we choose workers with the best reported transfers, etc. How much to consolidate is likely a workflow parameter.  Perhaps instead of sending files back to the manager, we can even have a handful of special workers (e.g. using --feature) that can act as stable storage to which files can be consolidated.
